## Python Pandas: Summarization

As always, let's begin by importing the pandas library under its pseudonym `pd`.

In [22]:
import pandas as pd

We'll continue working with the Netflix data that we worked with last time. Run the following code block to import that data again and merge the two datasets, as we discussed last time.

In [23]:
# read in the two CSVs
netflix_titles = pd.read_csv('netflix_titles.csv')
netflix_views = pd.read_csv('netflix_views.csv', encoding="latin-1")

# correct the foreign title issue from last week
netflix_views['Title'] = netflix_views['Title'].apply(lambda x: x.split(" // ")[0])

# turn the 'Hours Viewed' into a number format
netflix_views['Hours Viewed'] = netflix_views['Hours Viewed'].str.replace(",", "")
netflix_views['Hours Viewed'] = pd.to_numeric(netflix_views['Hours Viewed'])

# merge the data sets and see the first few rows
netflix = pd.merge(netflix_titles, netflix_views, left_on = "title", right_on = "Title")
netflix.iloc[:5]

show_id   type                             title  \
0      s7  Movie  My Little Pony: A New Generation   
1      s8  Movie                           Sankofa   
2     s10  Movie                      The Starling   
3     s13  Movie                      Je Suis Karl   
4     s14  Movie  Confessions of an Invisible Girl   

                        director  \
0  Robert Cullen, José Luis Ucha   
1                   Haile Gerima   
2                 Theodore Melfi   
3            Christian Schwochow   
4                  Bruno Garotti   

                                                cast  \
0  Vanessa Hudgens, Kimiko Glenn, James Marsden, ...   
1  Kofi Ghanaba, Oyafunmike Ogunlano, Alexandra D...   
2  Melissa McCarthy, Chris O'Dowd, Kevin Kline, T...   
3  Luna Wedler, Jannis Niewöhner, Milan Peschel, ...   
4  Klara Castanho, Lucca Picon, Júlia Gomes, Marc...   

                                             country          date_added  \
0                                                NaN  September 24, 2021   
1  United States, Ghana, Burkina Faso, United Kin...  September 24, 2021   
2                                      United States  September 24, 2021   
3                            Germany, Czech Republic  September 23, 2021   
4                                                NaN  September 22, 2021   

   release_year rating duration  \
0          2021     PG   91 min   
1          1993  TV-MA  125 min   
2          2021  PG-13  104 min   
3          2021  TV-MA  127 min   
4          2021  TV-PG   91 min   

                                          listed_in  \
0                          Children & Family Movies   
1  Dramas, Independent Movies, International Movies   
2                                  Comedies, Dramas   
3                      Dramas, International Movies   
4                Children & Family Movies, Comedies   

                                         description  \
0  Equestria's divided. But a bright-eyed hero be...   
1  On a photo shoot in Ghana, an American model s...   
2  A woman adjusting to life after a loss contend...   
3  After most of her family is murdered in a terr...   
4  When the clever but socially-awkward Tetê join...   

                              Title Available Globally? Release Date  \
0  My Little Pony: A New Generation                 Yes    9/24/2021   
1                           Sankofa                  No          NaN   
2                      The Starling                 Yes    9/24/2021   
3                      Je Suis Karl                 Yes    9/23/2021   
4  Confessions of an Invisible Girl                 Yes    9/22/2021   

   Hours Viewed  
0      15400000  
1        100000  
2       8200000  
3        300000  
4       5700000

## **Data Summaries**

Performing data summaries in pandas is essential for understanding the structure, trends, and key insights within a dataset. Let's begin with some simple functions to understand the size and scope of the data we're working with, without worrying about the information it actually contains.

### Getting info about your data

`.info()` is commonly one of the very first commands you run after loading your data:

In [3]:
netflix.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2895 entries, 0 to 2894
Data columns (total 16 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   show_id              2895 non-null   object
 1   type                 2895 non-null   object
 2   title                2895 non-null   object
 3   director             2735 non-null   object
 4   cast                 2728 non-null   object
 5   country              2649 non-null   object
 6   date_added           2895 non-null   object
 7   release_year         2895 non-null   int64 
 8   rating               2895 non-null   object
 9   duration             2892 non-null   object
 10  listed_in            2895 non-null   object
 11  description          2895 non-null   object
 12  Title                2895 non-null   object
 13  Available Globally?  2895 non-null   object
 14  Release Date         1170 non-null   object
 15  Hours Viewed         2895 non-null   int64 
dtypes: int

The `.info()` method provides the essential details about your dataset, such as the number of rows and columns, the number of non-null values, what type of data is in each column, and how much memory your DataFrame is using.

Notice in our Netflix dataset we have some obvious missing values in the `director`, `cast`, `country`, `duration`, and `Release Date` columns. There are a few ways you can handle these missing values, which we'll talk about later in the semester.

Another fast and useful attribute is `.shape`, which outputs just a tuple of (rows, columns):

In [4]:
netflix.shape

(2895, 16)

Note that `.shape` has no parentheses and is a simple tuple of format (rows, columns). So we have **2895 rows** and **16 columns** in our movies DataFrame.

You'll be going to `.shape` a lot when cleaning and transforming data. For example, you might filter some rows based on some criteria and then want to know quickly how many rows were removed.

Once we have this basic dataset information, we can start to think about the information the dataset contains.

`df.describe()` provides a quick statistical summary of numerical columns in a DataFrame, including metrics like count, mean, standard deviation, min, max, and quartiles. This function is useful for gaining an overview of data distributions, detecting outliers, and identifying potential issues such as missing or skewed values. By using `df.describe(include="all")`, you can also summarize non-numeric data, making it a versatile tool for exploratory data analysis.

In [25]:
netflix.describe()

release_year  Hours Viewed
count   2895.000000  2.895000e+03
mean    2014.864594  2.165941e+06
std        7.852936  5.219811e+06
min     1945.000000  1.000000e+05
25%     2014.000000  2.000000e+05
50%     2018.000000  5.000000e+05
75%     2019.000000  1.900000e+06
max     2021.000000  8.790000e+07

If you only need one specific piece of information from `df.describe()`, you can also extract that value on its own. For example, if you wanted to know the mean of the `Hours Viewed` column, you could type `netflix['Hours Viewed'].mean()`

`value_counts()` is a powerful method for analyzing the frequency of unique values in a Series. It helps identify common and rare categories, detect data imbalances, and spot potential data entry errors. By default, it returns counts in descending order, but you can customize it with parameters like `normalize=True` for proportions or `dropna=False` to include missing values. This function is especially useful for summarizing categorical data and gaining insights into distributions within a dataset.

In [27]:
netflix['rating'].value_counts()

rating
TV-MA     1104
TV-14      645
R          345
PG-13      232
TV-PG      197
PG         139
TV-Y        74
TV-Y7       72
TV-G        61
NR          11
G            9
NC-17        2
74 min       1
84 min       1
66 min       1
UR           1
Name: count, dtype: int64

### **Practice Questions:**

1. Calculate the range of values in the `release_year` column.

In [ ]:
# Answer here:
netflix['release_year'].max() - netflix['release_year'].min()

np.float64(38.0)

2. Which director(s) in this dataset has directed the 4th most films and/or TV shows?

In [32]:
# Answer here:
vc = netflix['director'].value_counts()
vc[vc == 9]

director
Rajiv Chilaka          9
Cathy Garcia-Molina    9
Name: count, dtype: int64

## **'Group By' Aggregation**

## Main Idea

Conceptually, the basic steps are as follows:

* **split** the data into several _groups_
    * e.g., split all Netflix offerings into the "TV show" and "movie" groups


* **apply** some function independently to each of the groups
    * e.g., average the hours viewed of all offerings in each group


* and then **merge** the results for all groups together
    * e.g., get back one Series with the average hours viewed indexed by type.

**Step 1: Split**

In [9]:
grps = netflix.groupby('type')
print(len(grps))

2


What does 'grps' look like?

In [10]:
print(type(grps))

<class 'pandas.core.groupby.generic.DataFrameGroupBy'>


This is not a DataFrame, or any object we have seen so far. However, in many respects, it acts as a _dictionary_.

In [11]:
# iterate over the keys and values of this groupby object
for key, group in grps:
    print('group name =', key)
    print(group[:5])

group name = Movie
  show_id   type                             title  \
0      s7  Movie  My Little Pony: A New Generation   
1      s8  Movie                           Sankofa   
2     s10  Movie                      The Starling   
3     s13  Movie                      Je Suis Karl   
4     s14  Movie  Confessions of an Invisible Girl   

                        director  \
0  Robert Cullen, José Luis Ucha   
1                   Haile Gerima   
2                 Theodore Melfi   
3            Christian Schwochow   
4                  Bruno Garotti   

                                                cast  \
0  Vanessa Hudgens, Kimiko Glenn, James Marsden, ...   
1  Kofi Ghanaba, Oyafunmike Ogunlano, Alexandra D...   
2  Melissa McCarthy, Chris O'Dowd, Kevin Kline, T...   
3  Luna Wedler, Jannis Niewöhner, Milan Peschel, ...   
4  Klara Castanho, Lucca Picon, Júlia Gomes, Marc...   

                                             country          date_added  \
0                         

_Summary_: We can think of the result of groupby() as a dictionary

* whose keys are the groups on which we split the data, and
* whose values are the DataFrame parts corresponding to those groups

**Step 2: Apply**

How do we find the average hours viewed for each 'type' grouping?

In [12]:
# find the mean hours viewed in each group
group_names = []
group_means = []
for name, group in grps:
    mean = group['Hours Viewed'].mean()
    print('mean for', name, 'is', mean)
    
    group_names.append(name)
    group_means.append(mean)
    

mean for Movie is 2103567.6061362824
mean for TV Show is 4066304.347826087


**Step 3: Merge**

How do we merge the means for all the groups into one Series?

In [13]:
# merge the means
pd.Series(group_means, index=group_names)

Movie      2.103568e+06
TV Show    4.066304e+06
dtype: float64

**Groupby:**

All of these steps can be compressed into one.

In [14]:
netflix.groupby('type')['Hours Viewed'].mean()

type
Movie      2.103568e+06
TV Show    4.066304e+06
Name: Hours Viewed, dtype: float64

Let us run through this groupby() in detail.

> netflix.groupby('type')['Hours Viewed'].mean()

* **netflix**: Obvious; run on the netflix DataFrame

* **groupby(['type'])**: Group together all Netflix offerings with the same type.

* **['Hours Viewed']**: Recall that for each of the groups, we get a DataFrame of offerings in that group. By specifying [column1, column2, ...], we _select_ those columns of the group DataFrames. Here, we select just the `Hours Viewed` column.

* **mean()**: Apply the mean() function on the selected columns (here, 'Hours Viewed') of each group's DataFrame.

For each of these three 'parts' of _groupby_ (not including the DataFrame), we can select multiple things at once. We are not limited to grouping by only one column, selecting only one column, or to performing only one function.

Here's an example of _grouping by_ multiple columns:

In [15]:
netflix.groupby(['type', 'director'])['Hours Viewed'].mean()

type     director                            
Movie    A. Raajdheep                             400000.0
         A. Salaam                                100000.0
         Aadish Keluskar                          200000.0
         Aamir Khan                              3400000.0
         Aaron Burns                              400000.0
                                                   ...    
TV Show  Sarawut Wichiensarn                      500000.0
         Shen Leping                              600000.0
         Stefano Lodovichi                        100000.0
         Su I-Hsuan, Chuang Shiang-an, Liu Yi     100000.0
         Tsutomu Mizushima                        100000.0
Name: Hours Viewed, Length: 2106, dtype: float64

The result of this groupby() is a DataFrame with a _hierarchical_ index with two levels. The first level of the index is type, and the second level is the director.

Here's an example of _selecting_ multiple columns:

In [16]:
netflix.groupby('type')[['Hours Viewed', 'release_year']].mean()

Hours Viewed  release_year
type                               
Movie    2.103568e+06   2014.827328
TV Show  4.066304e+06   2016.000000

And here's an example of _performing_ multiple functions:

In [17]:
netflix.groupby(['type'])['Hours Viewed'].agg(['mean', 'median'])

mean    median
type                           
Movie    2.103568e+06  500000.0
TV Show  4.066304e+06  650000.0

#### One last option: using custom functions

In [18]:
def range(s):
    """Given a series (called s), get the range"""
    return s.max() - s.min()

In [19]:
netflix.groupby('type')['Hours Viewed'].agg(range)

type
Movie      87800000
TV Show    77700000
Name: Hours Viewed, dtype: int64

### **Practice Problems:**

3. Name the three parts of a groupby function.

Answer here:
Group by columns, select columns, apply functions

4. Are movies or TV shows older on average?

In [33]:
# Answer here:
netflix.groupby('type')['release_year'].mean()

type
Movie      2014.827328
TV Show    2016.000000
Name: release_year, dtype: float64

5. Which movie/TV show `rating` has the highest average `Hours Viewed`? Which one has the greatest variability in `Hours Viewed`?

In [35]:
# Answer here:
netflix.groupby('rating')['Hours Viewed'].agg(['max', 'mean', 'std']).sort_values(by = 'mean', ascending = False)

max          mean           std
rating                                      
PG      56400000  6.155396e+06  1.054480e+07
UR       5700000  5.700000e+06           NaN
PG-13   87900000  3.927155e+06  8.098666e+06
R       86100000  3.563478e+06  7.175904e+06
TV-Y7   18800000  2.841667e+06  4.674828e+06
TV-Y    22300000  2.022973e+06  3.376208e+06
G        4800000  2.000000e+06  1.880824e+06
TV-G    24300000  1.644262e+06  3.424740e+06
TV-PG   40100000  1.504569e+06  3.503312e+06
TV-MA   24800000  1.429076e+06  2.631870e+06
TV-14   77800000  1.417209e+06  4.442856e+06
NR       2700000  6.090909e+05  7.816067e+05
NC-17     400000  2.500000e+05  2.121320e+05
74 min    200000  2.000000e+05           NaN
84 min    200000  2.000000e+05           NaN
66 min    200000  2.000000e+05           NaN